jupyterhub/spawner.py

```diff
@@ -685,12 +727,127 @@ def options_from_query(self, query_data):
         """
         return self.options_from_form(query_data)
 
+    apply_user_options = Union(
+        [Callable(), Dict()],
+        config=True,
+        default_value=None,
+        allow_none=True,
+        help="""
+            Apply inputs from user_options to the Spawner
+
+            Typically takes values in user_options, validates them, and updates Spawner attributes::
+
+                def apply_user_options(spawner, user_options):
+                    if "image" in user_options and isinstance(user_options["image"], str):
+                        spawner.image = user_options["image"]
+
+                c.Spawner.apply_user_options = apply_user_options
+
+            Default: do nothing
+
+            Typically a callalble which takes `(spawner: Spawner, user_options: dict)`,
+            but for simple cases this can be a dict mapping user option fields to Spawner attribute names,
+            e.g.::
+
+                c.Spawner.apply_user_options = {"image_input": "image"}
+
+            allows users to specify the image attribute, but not any others.
+
+            .. note::
+
+                Because `user_options` is user input
+                and may be set directly via the REST API,
+                no assumptions should be made on its structure or contents.
+                An empty dict should always be supported.
+                Make sure to validate any inputs before applying them!
+        
+            .. versionadded:: 5.3
+        
+                Prior to 5.3, applying user options must be done in `Spawner.start()`
+                or `Spawner.pre_spawn_hook()`.
+            """,
+    )
+
+    async def _run_apply_user_options(self, user_options):
+        """Run the apply_user_options hook
+
+        and turn errors into HTTP 400
+        """
+        r = None
+        try:
+            if isinstance(self.apply_user_options, dict):
+                r = self._apply_user_options_dict(user_options)
+            elif self.apply_user_options:
+                r = self.apply_user_options(self, user_options)
+            elif user_options:
+                keys = list(user_options)
+                self.log.warning(
+                    f"Received unhandled user_options for {self._log_name}: {', '.join(keys)}"
+                )
+            if isawaitable(r):
+                await r
+        except Exception as e:
+            # this may not be the users' fault...
+            # should we catch less?
+            # likely user errors are ValueError, TraitError, TypeError
+            self.log.exception("Exception applying user_options for %s", self._log_name)
+            raise web.HTTPError(400, f"Invalid user options: {e}")
```

Comment:
Do we have any guidelines in JupyterHub on which exceptions are exposed to users, to ensure internal information isn't leaked?

Reply:
We mention it [in authenticators](https://jupyterhub.readthedocs.io/en/stable/reference/authenticators.html#custom-error-messages), but the general pattern is that HTTPErrors are displayed to users. So if you want to present an error to a user, raising an HTTPError in your hook/Authenticator/Spawner method should let you display things to the user, while any other error should be logged and wrapped in a generic, appropriate HTTPError (usually 500).

You can also customize messages with a `jupyterhub[_html]_message` attribute on your exceptions, which affects what is displayed when things fail, but not the status code.

There have been issues where that passthrough doesn't always propagate correctly, but that's how it's supposed to work.

Reply:
Does this potentially break expectations then, since the content of a standard exception which would normally be internally logged is now automatically converted to one and exposed to the user?

Reply:
Yeah, good point. I'll switch this to do the passthrough, so only a specifically-raised HTTPError gets passed through, otherwise generic `HTTPError(400, "Invalid user options")`